engine/lib/phases/phase_drop_match_guards.ml

```diff
@@ -0,0 +1,200 @@
+(* This phase removes guards from pattern matchings. It rewrites
+   them using only pattern matchings without guards.
+   See #806 and the example in tests/guards. *)
+
+open! Prelude
+
+module%inlined_contents Make (F : Features.T) = struct
+  open Ast
+  module FA = F
+
+  module FB = struct
+    include F
+    include Features.Off.Match_guard
+  end
+
+  include
+    Phase_utils.MakeBase (F) (FB)
+      (struct
+        let phase_id = Diagnostics.Phase.DropMatchGuards
+      end)
+
+  module UA = Ast_utils.Make (F)
+  module UB = Ast_utils.Make (FB)
+
+  module Implem : ImplemT.T = struct
+    let metadata = metadata
+
+    module S = struct
+      include Features.SUBTYPE.Id
+    end
+
+    [%%inline_defs dmutability]
+
+    let rec dexpr' (span : span) (expr : A.expr') : B.expr' =
+      match expr with
+      | [%inline_arms "dexpr'.*" - Match] -> auto
+      | Match { scrutinee; arms } ->
+          let arms = transform_arms (dexpr scrutinee) (List.rev arms) [] in
```

Comment:
If you have `match <scrut> { ... if ... => ..., <other arms> }`, this would generate `match <scrut> { _ => match <scrut>} {...}}`, we should optimize that

Reply:
Yes, I also noticed that with tests. This phase can generate a lot of nested pattern matchings, some of them with only one Wildcard branch or other simplifiable situations. Do you think we should optimize in this phase (trying to simpler code in specific situations) or in another one (that would catch them and simplify)?

Reply:
I think we should do that in this phase: if a user writes a match like that in Rust, it's better not to rewrite it.
We should collapse matches only when a rewrite was done I think

Reply:
Most of the unnecessary `match ... { _ => ... }` seems to be gone :smiley: but if one appears at top-level, it seems like it's still there, e.g. https://hax-playground.cryspen.com/#fstar/9dc02d2817/gist=f490560c1ca7bb7a0e30f0e133fd8202
I guess you need to do something right there, when calling `transform_arms` in `dexpr'`, no?

Reply:
Good catch! Should be fixed now.